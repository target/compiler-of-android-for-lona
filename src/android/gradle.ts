export type AndroidConfig = {
  compileSdkVersion: number
  buildToolsVersion: string
  defaultConfig: {
    minSdkVersion: number
    targetSdkVersion: number
    versionCode: number
    versionName: string
  }
}

export type BuildScript = {
  plugins: string[]
  android: AndroidConfig
  dependencies: string[]
}

const createApplyPluginStatement = (plugin: string): string =>
  `apply plugin: "${plugin}"`

const createAndroidBlock = (android: AndroidConfig): string[] =>
  `android {
    compileSdkVersion ${android.compileSdkVersion}
    buildToolsVersion "${android.buildToolsVersion}"

    defaultConfig {
        minSdkVersion ${android.defaultConfig.minSdkVersion}
        targetSdkVersion ${android.defaultConfig.targetSdkVersion}
        versionCode ${android.defaultConfig.versionCode}
        versionName "${android.defaultConfig.versionName}"
    }
}`.split('\n')

const createDependencyStatement = (dependency: string): string =>
  `implementation "${dependency}"`

const createBlock = (identifier: string, statements: string[]): string[] =>
  statements.length === 0
    ? []
    : [
        `${identifier} {`,
        ...statements.map(statement => `    ${statement}`),
        `}`,
      ]

const addGroupSeparator = <Element, Group extends Array<Element>>(
  array: Group[],
  separator: Element
): Element[] => {
  return array
    .filter(inner => inner.length > 0)
    .flatMap((inner, index) =>
      index < array.length - 1 ? [...inner, separator] : inner
    )
}

export function createBuildScript(config: BuildScript): string {
  const applyPluginStatements = config.plugins.map(createApplyPluginStatement)
  const androidBlock = createAndroidBlock(config.android)
  const dependencyStatements = createBlock(
    'dependencies',
    config.dependencies.map(createDependencyStatement)
  )

  return addGroupSeparator(
    [applyPluginStatements, androidBlock, dependencyStatements],
    ''
  ).join('\n')
}

export const DEFAULT_PLUGINS = [
  'com.android.library',
  'kotlin-android',
  'kotlin-android-extensions',
]

export const DEFAULT_DEPENDENCIES = [
  'org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version',
  'androidx.appcompat:appcompat:1.1.0',
  'androidx.core:core-ktx:1.2.0',
]

export const DEFAULT_ANDROID_CONFIG: AndroidConfig = {
  compileSdkVersion: 29,
  buildToolsVersion: '29.0.2',
  defaultConfig: {
    minSdkVersion: 21,
    targetSdkVersion: 29,
    versionCode: 1,
    versionName: '1.0.0',
  },
}

export const DEFAULT_BUILD_CONFIG: BuildScript = {
  plugins: DEFAULT_PLUGINS,
  android: DEFAULT_ANDROID_CONFIG,
  dependencies: DEFAULT_DEPENDENCIES,
}
