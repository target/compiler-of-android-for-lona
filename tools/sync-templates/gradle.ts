import os from 'os'
import fs from 'fs-extra'
import path from 'path'
import { execSync } from 'child_process'

enum ProjectType {
  basic = 'basic',
}

enum DSL {
  groovy = 'groovy',
  kotlin = 'kotlin',
}

const commands = {
  gradleInit: ({
    type,
    dsl,
    projectName,
    projectDirectory,
  }: {
    type: ProjectType
    dsl: DSL
    projectName: string
    projectDirectory: string
  }): string => {
    return `gradle init --type ${type} --dsl ${dsl} --project-name ${projectName} --project-dir ${projectDirectory}`
  },
}

export default function sync() {
  // Create a temp directory
  const directoryPath = fs.mkdtempSync(path.join(os.tmpdir(), 'gradle-'))

  console.log(`Initializing gradle project in ${directoryPath}`)

  // Initialize a new gradle project
  execSync(
    commands.gradleInit({
      type: ProjectType.basic,
      dsl: DSL.groovy,
      projectName: 'temp-gradle-project',
      projectDirectory: directoryPath,
    })
  )

  const gradleWrapperTemplatePath = path.join(
    __dirname,
    '../../templates/android-studio/gradle/wrapper'
  )

  // Remove any old wrapper
  fs.removeSync(gradleWrapperTemplatePath)

  const wrapperFiles = ['gradle', 'gradlew', 'gradlew.bat']

  for (const file of wrapperFiles) {
    const sourcePath = path.join(directoryPath, file)
    const targetPath = path.join(gradleWrapperTemplatePath, file)

    console.log(`Copying gradle file ${sourcePath} to ${targetPath}`)

    // Copy new templates
    fs.copySync(sourcePath, targetPath)
  }

  // Clean up temp directory
  fs.removeSync(directoryPath)
}
