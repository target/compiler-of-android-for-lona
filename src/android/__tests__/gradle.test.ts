import { createBuildScript, DEFAULT_BUILD_CONFIG } from '../gradle'

describe('Android / Gradle', () => {
  test('creates a gradle file', () => {
    const buildScript = createBuildScript(DEFAULT_BUILD_CONFIG)

    expect(buildScript).toMatchSnapshot()
  })
})
