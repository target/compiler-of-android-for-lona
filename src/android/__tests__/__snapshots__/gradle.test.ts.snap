// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Android / Gradle creates a gradle file 1`] = `
"apply plugin: \\"com.android.library\\"
apply plugin: \\"kotlin-android\\"
apply plugin: \\"kotlin-android-extensions\\"

android {
    compileSdkVersion 29
    buildToolsVersion \\"29.0.2\\"

    defaultConfig {
        minSdkVersion 21
        targetSdkVersion 29
        versionCode 1
        versionName \\"1.0.0\\"
    }
}

dependencies {
    implementation \\"org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version\\"
    implementation \\"androidx.appcompat:appcompat:1.1.0\\"
    implementation \\"androidx.core:core-ktx:1.2.0\\"
}"
`;
